### ISSUE TITLE
Eliminate dependencies on C++ runtime DLL's

### ISSUE THREAD

**Issue by billziss-gh** [yogendersolanki91/winfsp#20]
Running `dumpbin /imports winfsp.net-x64.dll` on my system results in the following:

```
C:\...>dumpbin /imports winfsp.net-x64.dll
Microsoft (R) COFF/PE Dumper Version 14.00.24215.1
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file winfsp.net-x64.dll

File Type: DLL

  Section contains the following imports:

    VCRUNTIME140D.dll
             18000D0C8 Import Address Table
             1800354F0 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          13 __FrameUnwindFilter
                          25 __std_type_info_destroy_list
                           8 __C_specific_handler
                          3E memset
                          3C memcpy

    ucrtbased.dll
             18000D108 Import Address Table
             180035530 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          A4 _cexit
                          C1 _crt_at_quick_exit
                          C2 _crt_atexit
                         54A terminate
                         2B4 _register_onexit_function
                         172 _initialize_onexit_table
                         171 _initialize_narrow_environment
                          B6 _configure_narrow_argv
                          E5 _execute_onexit_table
                         3D5 abort
                         563 wcscpy_s
                         483 free
                         566 wcslen
                         4D7 malloc
                         2C1 _seh_filter_dll
                         175 _initterm_e
                         174 _initterm

    KERNEL32.dll
             18000D028 Import Address Table
             180035450 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                         592 UnhandledExceptionFilter
                         552 SetUnhandledExceptionFilter
                         2C5 GetStartupInfoW
                         4BC RtlVirtualUnwind
                         36A IsDebuggerPresent
                         370 IsProcessorFeaturePresent
                         561 Sleep
                         256 GetLastError
                         570 TerminateProcess
                         20F GetCurrentProcess
                         430 QueryPerformanceCounter
                         210 GetCurrentProcessId
                         214 GetCurrentThreadId
                         2DD GetSystemTimeAsFileTime
                         117 DisableThreadLibraryCalls
                         354 InitializeSListHead
                         4AE RtlCaptureContext
                         4B5 RtlLookupFunctionEntry
                         26D GetModuleHandleW

    ADVAPI32.dll
             18000D000 Import Address Table
             180035428 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                         15E GetSecurityDescriptorLength
                          80 ConvertStringSecurityDescriptorToSecurityDescriptorA
                          81 ConvertStringSecurityDescriptorToSecurityDescriptorW
                          78 ConvertSecurityDescriptorToStringSecurityDescriptorA

    winfsp-x64.dll
             18000D198 Import Address Table
             1800355C0 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          33 FspFileSystemStartDispatcher
                          31 FspFileSystemSetMountPoint
                          3B FspNtStatusFromWin32
                          2F FspFileSystemResolveReparsePoints
                          10 FspFileSystemAddDirInfo
                          13 FspFileSystemCreate

    mscoree.dll
             18000D0F8 Import Address Table
             180035520 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          76 _CorDllMain

  Summary

        1000 .data
        1000 .gfids
        1000 .nep
        1000 .pdata
       29000 .rdata
        1000 .reloc
        1000 .rsrc
        B000 .text
```

Notice that `winfsp.net-x64.dll` depends on `VCRUNTIME140D.dll` and `ucrtbased.dll`.

This unfortunately creates a dependency of WinFsp on the MSVC runtime. Recall that WinFsp does not currently have such a dependency and I would much prefer if we did not introduce one. I am investigating if it is possible to build the .NET DLL without this dependency.

There are many reasons that I would rather not have WinFsp depend on the Microsoft runtimes. Here are some:

- There are multiple versions of these DLL's. All too often a new Visual Studio version contains a new DLL. Even the same VS version usually contains a Debug and a Release DLL.

- It would require us to produce multiple versions of our DLL's to target all the different runtime combinations: old vs new (ucrt-based) runtime, debug vs release, etc.

- The WinFsp installer is currently very simple and cannot handle installation of MSVC runtimes.

- It would be very confusing for users to choose the right DLL to link against.

**Comment by billziss-gh** [comment]
This may not actually be possible with C++/CLI. Bummer :(

With a bit of work (using `shared/minimal.h`, eliminating `wcscpy_s`, etc.) I got:
```
error LNK2001: unresolved external symbol "?.cctor@@$$FYMXXZ" (?.cctor@@$$FYMXXZ)
```

Interesting link:

https://groups.google.com/d/msg/microsoft.public.dotnet.languages.vc/TTE3xbsDxt0/g9XaSy3A_mIJ

**Comment by yogendersolanki91** [comment]
I think VCRUNTIME140D.dll is not needed for Release build.

**Comment by billziss-gh** [comment]
> I think VCRUNTIME140D.dll is not needed for Release build.

That is correct. For release builds it needs `VCRUNTIME140.dll` and other parts of the ucrt (`api-*.dll`):

```
C:\...\winfsp\build\VStudio\build\Release>dumpbin /imports winfsp.net-x64.dll
Microsoft (R) COFF/PE Dumper Version 14.00.24215.1
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file winfsp.net-x64.dll

File Type: DLL

  Section contains the following imports:

    VCRUNTIME140.dll
             1800060A8 Import Address Table
             18000FB00 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          3E memset
                          25 __std_type_info_destroy_list
                          13 __FrameUnwindFilter
                           8 __C_specific_handler

    api-ms-win-crt-runtime-l1-1-0.dll
             1800060E8 Import Address Table
             18000FB40 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          1E _crt_atexit
                          22 _execute_onexit_table
                          3C _register_onexit_function
                          34 _initialize_onexit_table
                          16 _cexit
                          18 _configure_narrow_argv
                          3F _seh_filter_dll
                          37 _initterm_e
                          36 _initterm
                          54 abort
                          33 _initialize_narrow_environment

    KERNEL32.dll
             180006028 Import Address Table
             18000FA80 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                         4B5 RtlLookupFunctionEntry
                         4BC RtlVirtualUnwind
                         36A IsDebuggerPresent
                         552 SetUnhandledExceptionFilter
                         4AE RtlCaptureContext
                         561 Sleep
                         256 GetLastError
                         370 IsProcessorFeaturePresent
                         354 InitializeSListHead
                         430 QueryPerformanceCounter
                         210 GetCurrentProcessId
                         214 GetCurrentThreadId
                         2DD GetSystemTimeAsFileTime
                         117 DisableThreadLibraryCalls
                         592 UnhandledExceptionFilter

    ADVAPI32.dll
             180006000 Import Address Table
             18000FA58 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          78 ConvertSecurityDescriptorToStringSecurityDescriptorA
                          80 ConvertStringSecurityDescriptorToSecurityDescriptorA
                          81 ConvertStringSecurityDescriptorToSecurityDescriptorW
                         15E GetSecurityDescriptorLength

    winfsp-x64.dll
             180006168 Import Address Table
             18000FBC0 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          33 FspFileSystemStartDispatcher
                          31 FspFileSystemSetMountPoint
                          13 FspFileSystemCreate
                          2F FspFileSystemResolveReparsePoints
                          10 FspFileSystemAddDirInfo
                          3B FspNtStatusFromWin32

    api-ms-win-crt-heap-l1-1-0.dll
             1800060D0 Import Address Table
             18000FB28 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          19 malloc
                          18 free

    api-ms-win-crt-string-l1-1-0.dll
             180006148 Import Address Table
             18000FBA0 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          A1 wcscpy_s

    mscoree.dll
             180006158 Import Address Table
             18000FBB0 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                          76 _CorDllMain

  Summary

        1000 .data
        1000 .gfids
        1000 .nep
        1000 .pdata
        B000 .rdata
        1000 .reloc
        1000 .rsrc
        4000 .text
```

I do not think it is possible to get rid of this dependency for C++/CLI. [For plain C++ we could link with the static CRT, but this is not possible with C++/CLI.]

**Comment by yogendersolanki91** [comment]
I think that is not even possible with C# code. These file are part of CLR and .NET framework. Even for non C++/CLI code (simple C# code) these files are required. Let us see what we can do about this.

Thanks

**Comment by billziss-gh** [comment]
I have been experimenting over the weekend to find a solution for C++/CLI. Unfortunately I do not have a good answer.

About the only solution that I could think of is to eliminate as many dependencies as possible and implement the remaining CRT symbols. For example this is what I do for C with the simple file [`shared/minimal.h`](https://github.com/yogendersolanki91/winfsp/blob/master/src/shared/minimal.h).

For C++ things are far more difficult. While it is easy to implement custom `new` / `delete` and eliminate `memcpy`, etc. there are other aspects that are far more difficult, e.g. exceptions and unwind support, etc. Nevertheless I have been successful in the past in doing it when I had to meet some strict requirements (on a different platform).

For C++/CLI things are probably impossible. It is quite likely that the CLR has certain expectations from the C++ runtime that could not be met by us simply reimplementing some symbols.

Unfortunately this makes C++/CLI undesirable as a solution of choice for such a layer that is incorporated within WinFsp. I really would like to keep WinFsp "lean and mean" and with no requirements other than the base OS.

My suggestion would be to keep this layer separate from the core WinFsp. On the plus side this also gives you complete freedom to design it in any way you want and you do not have to assign any copyright to me. And I will absolutely continue to help of course.

I have also been experimenting with a C# layer over the weekend. Progress has been slow but steady, because my C# is very rusty as discussed earlier. I have found that C# DLL's only require DLL's from the base OS. You are welcome to help there if you are interested.
